R/transform_quantile_to_pmf.R

```diff
@@ -0,0 +1,264 @@
+  model_out_tbl <- val_transform_q_to_p_args(model_out_tbl, bin_endpoints)
+  do_transform_q_to_p(model_out_tbl, bin_endpoints, tail_dist)
+}
+
+#' Perform the operation of transforming quantile forecasts to pmf forecasts
+#' @noRd
+do_transform_q_to_p <- function(model_out_tbl, bin_endpoints, tail_dist) {
+  task_id_cols <- get_task_id_cols(model_out_tbl)
+  join_cols <- task_id_cols[task_id_cols %in% colnames(bin_endpoints)]
+
+  # get one estimated cdf per model and prediction task
+  model_out_tbl <- model_out_tbl |>
+    dplyr::group_by(dplyr::across(dplyr::all_of(c("model_id", task_id_cols)))) |>
+    dplyr::summarize(
+      p_fn = list(distfromq::make_p_fn(
+        ps = as.numeric(.data[["output_type_id"]]),
+        qs = as.numeric(.data[["value"]]),
+        tail_dist = tail_dist
+      ))
+    )
+
+  # join with bin endpoints
+  #
+  # left join because we only need rows from model_out_tbl (e.g. bin_endpoints
+  # may have more locations than model_out_tbl)
+  #
+  # many-to-many relationship because:
+  # - bin_endpoints may have a subset of task id columns, does not have model_id
+  # - bin_endpoints has many output_type_ids
+  # dplyr complains if we try to join on an empty character vector,
+  # so we cross_join in that case
+  if (length(join_cols) > 0) {
+    model_out_tbl <- model_out_tbl |>
+      dplyr::left_join(
+        bin_endpoints,
+        by = join_cols,
+        relationship = "many-to-many"
+      )
+  } else {
+    model_out_tbl <- model_out_tbl |>
+      dplyr::cross_join(
+        bin_endpoints
+      )
+  }
+
+  # compute class probabilities, clean up included columns
+  model_out_tbl <- model_out_tbl |>
+    dplyr::mutate(
+      value = purrr::pmap_dbl(
+        list(.data[["p_fn"]], .data[["lower"]], .data[["upper"]]),
+        function(p_fn, lower, upper) {
+          p_fn(upper) - p_fn(lower)
+        }
+      )
+    ) |>
+    dplyr::mutate(
+      output_type = "pmf",
+      .before = "output_type_id"
+    ) |>
+    dplyr::select(
+      -dplyr::all_of(c("lower", "upper", "p_fn"))
+    )
+
+  return(model_out_tbl)
+}
+
+#' Get task id columns
+#' @noRd
+get_task_id_cols <- function(model_out_tbl) {
+  model_out_cols <- colnames(model_out_tbl)
+  non_task_cols <- c("model_id", "output_type", "output_type_id", "value")
+  task_id_cols <- model_out_cols[!model_out_cols %in% non_task_cols]
+
+  return(task_id_cols)
+}
+
+#' Do validation of arguments to transform_quantile_to_pmf
+#' @noRd
+val_transform_q_to_p_args <- function(model_out_tbl, bin_endpoints) {
+  if (!inherits(model_out_tbl, "model_out_tbl")) {
+    model_out_tbl <- hubUtils::as_model_out_tbl(model_out_tbl)
+  }
+  unique_output_types <- unique(model_out_tbl$output_type)
+  if (! "quantile" %in% unique_output_types) {
+    cli::cli_abort("{.arg model_out_tbl} must contain predictions with output type 'quantile'.")
+  }
+  model_out_tbl <- model_out_tbl |> dplyr::filter(.data[["output_type"]] == "quantile")
```

Comment:
If we're allowing the `model_out_tbl` to have non-quantile output types but are filtering those out, we should warn the user and/or note this in the function description (I prefer to do both).

Reply:
Thanks for the suggestion!

In [244ef6d](https://github.com/reichlab/idforecastutils/pull/11/commits/244ef6d91c8abf1a19d17ed7419291ee8d2a231b) I updated the documentation to clarify that the returned object has only the new-created pmf forecasts.  I was thinking this would address your concern by more generally clarifying that everything in the input `model_out_tbl` is discarded.  Does that make sense?

Reply:
yes it makes sense to me!